src/Microsoft.AspNetCore.Razor.VSCode/src/RazorDocumentSynchronizer.ts

```diff
@@ -24,164 +25,257 @@ export class RazorDocumentSynchronizer {
     }
 
     public register() {
-        const changeRegistration = vscode.workspace.onDidChangeTextDocument((args) => {
-            if (args.document.languageId !== RazorLanguage.id) {
-                return;
-            }
-
-            const uriPath = getUriPath(args.document.uri);
-            const context = this.synchronizations[uriPath];
-
-            if (context && args.document.version >= context.documentVersion) {
-                if (this.logger.verboseEnabled) {
-                    this.logger.logVerbose(
-                        `${context.logIdentifier} - Notify Success: Host document updated to equivalent version.`);
-                }
-                context.synchronized(true);
-            }
-        });
-
-        return changeRegistration;
+        return vscode.workspace.onDidChangeTextDocument(event => this.textDocumentChanged(event));
     }
 
     public async trySynchronize(
         hostDocument: vscode.TextDocument,
         projectedDocument: IProjectedDocument,
-        toVersion: number) {
-        const logIdentifier = this.synchronizationIdentifier++;
+        toHostDocumentVersion: number,
+        token: vscode.CancellationToken) {
 
+        const logId = ++this.synchronizationIdentifier;
+
+        const documentKey = getUriPath(projectedDocument.uri);
         if (this.logger.verboseEnabled) {
-            this.logger.logVerbose(`${logIdentifier} - Synchronizing '${getUriPath(projectedDocument.uri)}' ` +
-                `currently at ${projectedDocument.hostDocumentSyncVersion} to version '${toVersion}'. ` +
-                `Current host document version: '${hostDocument.version}'`);
+            this.logger.logVerbose(
+                `${logId} - Synchronizing '${documentKey}':
+    Currently at ${projectedDocument.hostDocumentSyncVersion}, synchronizing to version '${toHostDocumentVersion}'.
+    Current host document version: '${hostDocument.version}'
+    Current projected document version: '${projectedDocument.projectedDocumentSyncVersion}'`);
         }
 
-        if (projectedDocument.hostDocumentSyncVersion === hostDocument.version) {
+        if (hostDocument.version !== toHostDocumentVersion) {
```

Comment:
This check feels odd to me. Maybe I'm just misreading it - hope you can clarify.

The caller has asked for `hostDocument` to be synchronized to version `toHostDocumentVersion`. So the caller probably doesn't expect for `hostDocument`'s version already to equal `toHostDocumentVersion` (otherwise they wouldn't be calling this). So why do we log a warning to say the version doesn't already match?

Maybe some renaming of args would clarify the intent.

Reply:
Oh wait I think I understand now. This method intends to synchronize the *projected* document with the host document. Your check is to validate that the caller is correct about the current version of the host document, not the projected document that we are updating. I don't 100% know why we need to check that (instead of dropping the `toHostDocumentVersion` param, and just always updating the projected document to match `hostDocument`) but it's fine if you have some reason!

Maybe to clarify, this method could be renamed to `trySynchronizeProjectedDocument`, and the `toHostDocumentVersion` param could be renamed to `expectedHostDocumentVersion`.

Reply:
Hmmm, love the new naming (and you're dead on for what it's trying to accomplish), i'll definitely do that.

Reply:
Btw, how bad would it be if we nuked the line length restriction in TypeScript? There are so many places where I feel like the restriction is more harmful then beneficial